File: Java/Ema/Src/main/java/interface/com/thomsonreuters/ema/access/OmmState.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright Thomson Reuters 2015. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.thomsonreuters.ema.access;

/**
 * OmmState represents State information in Omm.
 * <p>OmmState is used to represent state of item, item group and service.</p>
 * <p>OmmState encapsulates stream state, data state, status code and status text information.</p>
 * 
 * <p>OmmState is a read only class.
 * <br>OmmState is used for extraction of OmmState info only.</p>
 */
public interface OmmState extends Data
{
	/** 
	 * StreamState represents item stream state.
	 */
	public static class StreamState
	{
		/**
		 * Indicates the stream is opened and will incur interest after the final refresh.
		 */
		public final static int OPEN			 = 1;
	
		/**
		 * Indicates the item will not incur interest after	the final refresh.
		 */
		public final static int NON_STREAMING	 = 2;
	
		/**
		 * Indicates the stream is closed, typically unobtainable or 
		 * identity indeterminable due to a comms outage.
		 * The item may be available in the future.
		 */
		public final static int	CLOSED_RECOVER	 = 3;
	
		/**
		 * Indicates the stream is closed.
		 */
		public final static int CLOSED			 = 4;
	
		/**
		 * Indicates the stream is closed and has been renamed.
		 * The stream is available with another name.
		 * This stream state is only valid for refresh messages.
		 * The new item name is in the Name get accessor methods.
		 */
		public final static int CLOSED_REDIRECTED = 5;
	}
	
	/** 
	 * DataState represents item data state.
	 */
	public static class DataState
	{
		/**
		 * Indicates the health of the data item did not change.
		 */
		public final static int NO_CHANGE	= 0;
	
		/**
		 * Indicates the entire data item is healthy.
		 */
		public final static int OK			= 1;
	
		/**
		 * Indicates the health of some or all of the item's data is stale or unknown.
		 */
		public final static int SUSPECT		= 2;
	}
	
	/** 
	 * StatusCode represents status code.
	 */
	public static class  StatusCode
	{
		/** None */
		public final static int NONE						= 0;

		/** Not Found */
		public final static int NOT_FOUND					= 1;
		
		/** Timeout */
		public final static int TIMEOUT						= 2;
		
		/** Not Authorized */
		public final static int NOT_AUTHORIZED				= 3;
		
		/** Invalid Argument */
		public final static int INVALID_ARGUMENT			= 4;
		
		/** Usage Error */
		public final static int USAGE_ERROR					= 5;
		
		/** Pre-empted */
		public final static int PREEMPTED					= 6;
		
		/** Just In Time Filtering Started */
		public final static int JUST_IN_TIME_CONFLATION_STARTED	= 7;
		
		/** Tick By Tick Resumed */
		public final static int TICK_BY_TICK_RESUMED		= 8;
		
		/** Fail-over Started */
		public final static int FAILOVER_STARTED			= 9;
		
		/** Fail-over Completed */
		public final static int FAILOVER_COMPLETED			= 10;
		
		/** Gap Detected */
		public final static int GAP_DETECTED				= 11;
		
		/** No Resources */
		public final static int NO_RESOURCES				= 12;
		
		/** Too Many Items */
		public final static int TOO_MANY_ITEMS				= 13;
		
		/** Already Open */
		public final static int ALREADY_OPEN				= 14;
		
		/** Source Unknown */
		public final static int SOURCE_UNKNOWN				= 15;
		
		/** Not Open */
		public final static int NOT_OPEN					= 16;
		
		/** Non Updating Item */
		public final static int NON_UPDATING_ITEM			= 19;
		
		/** Unsupported View Type */
		public final static int UNSUPPORTED_VIEW_TYPE		= 20;
		
		/** Invalid View */
		public final static int INVALID_VIEW				= 21;
		
		/** Full View Provided */
		public final static int FULL_VIEW_PROVIDED			= 22;
		
		/** Unable To Request As Batch */
		public final static int UNABLE_TO_REQUEST_AS_BATCH		= 23;
		
		/** Request does not support batch or view */
		public final static int NO_BATCH_VIEW_SUPPORT_IN_REQ	= 26;
		
		/** Exceeded maximum number of mounts per user */
		public final static int EXCEEDED_MAX_MOUNTS_PER_USER	= 27;
		
		/** Internal error from sender */
		public final static int ERROR							= 28;
		
		/** Connection to DACS down, users are not allowed to connect */
		public final static int DACS_DOWN						= 29;
		
		/** User unknown to permissioning system, it could be DACS, AAA or EED */
		public final static int USER_UNKNOWN_TO_PERM_SYS		= 30;
		
		/** Maximum logins reached */
		public final static int DACS_MAX_LOGINS_REACHED			= 31;
		
		/** User is not allowed to use application */
		public final static int DACS_USER_ACCESS_TO_APP_DENIED	= 32;
		
		public final static int INVALID_FORMED_MSG				= 256;
		public final static int CHANNEL_UNAVAILABLE				= 257;
		public final static int SERVICE_UNAVAILABLE				= 258;
		public final static int SERVICE_DOWN					= 259;
		public final static int SERVICE_NOT_ACCEPTING_REQUESTS  = 260;
		public final static int LOGIN_CLOSED					= 261;
		public final static int DIRECTORY_CLOSED				= 262;
		public final static int ITEM_NOT_FOUND					= 263;
		public final static int DICTIONARY_UNAVAILABLE			= 264;
		public final static int FIELD_ID_NOT_FOUND_DICTIONARY_UNAVAILABLE = 265;
		public final static int ITEM_REQUEST_TIMEOUT			= 266;
	}
	
	/**
	 * Returns the StreamState value as a string format.
	 * 
	 * @return string representation of this object's StreamState
	 */
	public String streamStateAsString();
		
	/**
	 * Returns the DataState value as a string format.
	 * 
	 * @return string representation of this object's DataState
	*/
	public String dataStateAsString();
		
	/**
	 * Returns the StatusCode value as a string format.
	 * 
	 * @return string representation of this object's StatusCode
	 */
	public String statusCodeAsString();
		
	/**
	 * Returns StreamState.
	 * 
	 * @return value of StreamState
	 */
	public int streamState();
	
	/**
	 * Returns DataState.
	 * 
	 * @return value of DataState
	*/
	public int dataState();
	
	/**
	 * Returns StatusCode.
	 * 
	 * @return value of StatusCode
	 */
	public int statusCode();
	
	/**
	 * Returns StatusText.
	 * 
	 * @return String containing status text information
	 */
	public String statusText();
}

File: CSharp/Ema/Src/Core/Access/OmmState.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2023 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Codec;
using System;

namespace LSEG.Ema.Access;

/// <summary>
/// OmmState represents State information in Omm.
/// </summary>
/// <remarks>
/// OmmState is used to represent state of item, item group and service.<br/>
/// OmmState encapsulates stream state, data state, status code and status text information.<br/>
/// OmmState is a read only class.<br/>
/// This class is used for extraction of OmmState info only.<br/>
/// All methods in this class are single threaded.<br/>
/// </remarks>
/// <seealso cref="Data"/>
/// <seealso cref="EmaBuffer"/>
public sealed class OmmState : Data
{

    #region Public members

    /// <summary>
    /// StreamState represents item stream state.
    /// </summary>
    public static class StreamStates
    {
        /// <summary>
        /// Indicates the stream is opened and will incur interest after the final refresh.
        /// </summary>
        public const int OPEN = 1;

        /// <summary>
        /// Indicates the item will not incur interest after	the final refresh.
        /// </summary>
        public const int NON_STREAMING = 2;

        /// <summary>
        /// Indicates the stream is closed, typically unobtainable or
        /// identity indeterminable due to a comms outage.<br/>
        /// The item may be available in the future.
        /// </summary>
        public const int CLOSED_RECOVER = 3;

        /// <summary>
        /// Indicates the stream is closed.
        /// </summary>
        public const int CLOSED = 4;

        /// <summary>
        /// Indicates the stream is closed and has been renamed.<br/>
        /// The stream is available with another name.<br/>
        /// This stream state is only valid for refresh messages.<br/>
        /// The new item name is in the Name get accessor methods.
        /// </summary>
        public const int CLOSED_REDIRECTED = 5;
    }

    /// <summary>
    /// DataState represents item data state.
    /// </summary>
    public static class DataStates
    {
        /// <summary>
        /// Indicates the health of the data item did not change.
        /// </summary>
        public const int NO_CHANGE = 0;

        /// <summary>
        /// Indicates the entire data item is healthy.
        /// </summary>
        public const int OK = 1;

        /// <summary>
        /// Indicates the health of some or all of the item's data is stale or unknown.
        /// </summary>
        public const int SUSPECT = 2;
    }

    /// <summary>
    /// StatusCode represents status code.
    /// </summary>
    public static class StatusCodes
    {
        /// <summary> None </summary>
        public const int NONE = 0;

        /// <summary> Not Found </summary>
        public const int NOT_FOUND = 1;

        /// <summary> Timeout </summary>
        public const int TIMEOUT = 2;

        /// <summary> Not Authorized </summary>
        public const int NOT_AUTHORIZED = 3;

        /// <summary> Invalid Argument </summary>
        public const int INVALID_ARGUMENT = 4;

        /// <summary> Usage Error </summary>
        public const int USAGE_ERROR = 5;

        /// <summary> Pre-empted </summary>
        public const int PREEMPTED = 6;

        /// <summary> Just In Time Filtering Started </summary>
        public const int JUST_IN_TIME_CONFLATION_STARTED = 7;

        /// <summary> Tick By Tick Resumed </summary>
        public const int TICK_BY_TICK_RESUMED = 8;

        /// <summary> Fail-over Started </summary>
        public const int FAILOVER_STARTED = 9;

        /// <summary> Fail-over Completed </summary>
        public const int FAILOVER_COMPLETED = 10;

        /// <summary> Gap Detected </summary>
        public const int GAP_DETECTED = 11;

        /// <summary> No Resources </summary>
        public const int NO_RESOURCES = 12;

        /// <summary> Too Many Items </summary>
        public const int TOO_MANY_ITEMS = 13;

        /// <summary> Already Open </summary>
        public const int ALREADY_OPEN = 14;

        /// <summary> Source Unknown </summary>
        public const int SOURCE_UNKNOWN = 15;

        /// <summary> Not Open </summary>
        public const int NOT_OPEN = 16;

        /// <summary> Non Updating Item </summary>
        public const int NON_UPDATING_ITEM = 19;

        /// <summary> Unsupported View Type </summary>
        public const int UNSUPPORTED_VIEW_TYPE = 20;

        /// <summary> Invalid View </summary>
        public const int INVALID_VIEW = 21;

        /// <summary> Full View Provided </summary>
        public const int FULL_VIEW_PROVIDED = 22;

        /// <summary> Unable To Request As Batch </summary>
        public const int UNABLE_TO_REQUEST_AS_BATCH = 23;

        /// <summary> Request does not support batch or view </summary>
        public const int NO_BATCH_VIEW_SUPPORT_IN_REQ = 26;

        /// <summary> Exceeded maximum number of mounts per user </summary>
        public const int EXCEEDED_MAX_MOUNTS_PER_USER = 27;

        /// <summary> Internal error from sender </summary>
        public const int ERROR = 28;

        /// <summary> Connection to DACS down, users are not allowed to connect </summary>
        public const int DACS_DOWN = 29;

        /// <summary> User unknown to permissioning system, it could be DACS, AAA or EED </summary>
        public const int USER_UNKNOWN_TO_PERM_SYS = 30;

        /// <summary> Maximum logins reached </summary>
        public const int DACS_MAX_LOGINS_REACHED = 31;

        /// <summary> User is not allowed to use application </summary>
        public const int DACS_USER_ACCESS_TO_APP_DENIED = 32;

        /// <summary> Content is intended to fill a recognized gap </summary>
        public const int GAP_FILL = 34;

        /// <summary> Application Authorization Failed </summary>
        public const int APP_AUTHORIZATION_FAILED = 35;
    }

    /// <summary>
    /// Gets stream state.
    /// </summary>
    public int StreamState { get => m_State.StreamState(); }

    /// <summary>
    /// Gets data state.
    /// </summary>
    public int DataState { get => m_State.DataState(); }

    /// <summary>
    /// Gets status code.
    /// </summary>
    public int StatusCode { get => m_State.Code(); }

    /// <summary>
    /// Returns the StreamState value as a string format.
    /// </summary>
    /// <returns>string representation of this object's StreamState</returns>
    public string StreamStateAsString()
    {
        return StreamState switch
        {
            StreamStates.OPEN => OPEN_STRING,
            StreamStates.NON_STREAMING => NONSTREAMING_STRING,
            StreamStates.CLOSED => CLOSED_STRING,
            StreamStates.CLOSED_RECOVER => CLOSEDRECOVER_STRING,
            StreamStates.CLOSED_REDIRECTED => CLOSEDREDIRECTED_STRING,
            _ => DEFAULTSS_STRING + StreamState
        };
    }

    /// <summary>
    /// Returns the DataState value as a string format.
    /// </summary>
    /// <returns>string representation of this object's DataState</returns>
    public string DataStateAsString()
    {
        return DataState switch
        {
            DataStates.NO_CHANGE => NOCHANGE_STRING,
            DataStates.OK => OK_STRING,
            DataStates.SUSPECT => SUSPECT_STRING,
            _ => DEFAULTDS_STRING + DataState
        };
    }

    /// <summary>
    /// Returns the StatusCode value as a string format.
    /// </summary>
    /// <returns>string representation of this object's StatusCode</returns>
    public string StatusCodeAsString()
    {
        return StatusCode switch
        {
            StatusCodes.NONE => NONE_STRING,
            StatusCodes.NOT_FOUND => NOTFOUND_STRING,
            StatusCodes.TIMEOUT => TIMEOUT_STRING,
            StatusCodes.NOT_AUTHORIZED => NOTAUTHORIZED_STRING,
            StatusCodes.INVALID_ARGUMENT => INVALIDARGUMENT_STRING,
            StatusCodes.USAGE_ERROR => USAGEERROR_STRING,
            StatusCodes.PREEMPTED => PREEMPTED_STRING,
            StatusCodes.JUST_IN_TIME_CONFLATION_STARTED => JUSTINTIMECONFLATIONSTARTED_STRING,
            StatusCodes.TICK_BY_TICK_RESUMED => TICKBYTICKRESUMED_STRING,
            StatusCodes.FAILOVER_STARTED => FAILOVERSTARTED_STRING,
            StatusCodes.FAILOVER_COMPLETED => FAILOVERCOMPLETED_STRING,
            StatusCodes.GAP_DETECTED => GAPDETECTED_STRING,
            StatusCodes.NO_RESOURCES => NORESOURCES_STRING,
            StatusCodes.TOO_MANY_ITEMS => TOOMANYITEMS_STRING,
            StatusCodes.ALREADY_OPEN => ALREADYOPEN_STRING,
            StatusCodes.SOURCE_UNKNOWN => SOURCEUNKNOWN_STRING,
            StatusCodes.NOT_OPEN => NOTOPEN_STRING,
            StatusCodes.NON_UPDATING_ITEM => NONUPDATINGITEM_STRING,
            StatusCodes.UNSUPPORTED_VIEW_TYPE => UNSUPPORTEDVIEWTYPE_STRING,
            StatusCodes.INVALID_VIEW => INVALIDVIEW_STRING,
            StatusCodes.FULL_VIEW_PROVIDED => FULLVIEWPROVIDED_STRING,
            StatusCodes.UNABLE_TO_REQUEST_AS_BATCH => UNABLETOREQUESTASBATCH_STRING,
            StatusCodes.NO_BATCH_VIEW_SUPPORT_IN_REQ => NOBATCHVIEWSUPPORTINREQ_STRING,
            StatusCodes.EXCEEDED_MAX_MOUNTS_PER_USER => EXCEEDEDMAXMOUNTSPERUSER_STRING,
            StatusCodes.ERROR => ERROR_STRING,
            StatusCodes.DACS_DOWN => DACSDOWN_STRING,
            StatusCodes.USER_UNKNOWN_TO_PERM_SYS => USERUNKNOWNTOPERMSYS_STRING,
            StatusCodes.DACS_MAX_LOGINS_REACHED => DACSMAXLOGINSREACHED_STRING,
            StatusCodes.DACS_USER_ACCESS_TO_APP_DENIED => DACSUSERACCESSTOAPPDENIED_STRING,
            StatusCodes.GAP_FILL => GAPFILL_STRING,
            StatusCodes.APP_AUTHORIZATION_FAILED => APPAUTHORIZATIONFAILED_STRING,
            _ => DEFAULTSC_STRING + StatusCode
        };
    }

    /// <summary>
    /// Gets status text.
    /// </summary>
    public string StatusText
    {
        get
        {
            if (m_State.Text().Length == 0)
                return string.Empty;
            else
                return m_State.Text().ToString();
        }
    }

    /// <summary>
    /// Provides string representation of the current instance
    /// </summary>
    /// <returns>string representing current <see cref="OmmState"/> object.</returns>
    public override string ToString()
    {
        if (DataCode.BLANK == Code)
            return BLANK_STRING;
        else
        {
            m_ToString.Clear();
            m_ToString.Append(Eta.Codec.StreamStates.Info(m_State.StreamState()))
                .Append(" / ")
                .Append(Eta.Codec.DataStates.Info(m_State.DataState()))
                .Append(" / ")
                .Append(Eta.Codec.StateCodes.Info(m_State.Code()))
                .Append(" / '")
                .Append(StatusText)
                .Append("'");

            return m_ToString.ToString();
        }
    }

    #endregion

    #region Implementation details

    internal OmmState() 
    { 
        ReturnToPoolInternal = () => m_objectManager.ReturnToPool(this);
        DecodePrimitiveType = DecodeOmmState;
        m_dataType = Access.DataType.DataTypes.STATE;
    }

    internal CodecReturnCode DecodeOmmState(DecodeIterator dIter)
    {

        if (CodecReturnCode.SUCCESS == m_State.Decode(dIter))
            Code = DataCode.NO_CODE;
        else
        {
            Code = DataCode.BLANK;
            m_State.Clear();
        }

        return CodecReturnCode.SUCCESS;
    }

    internal void Decode(Eta.Codec.State rsslState)
    {
        if (rsslState != null)
        {
            Code = DataCode.NO_CODE;

            m_State.StreamState(rsslState.StreamState());
            m_State.DataState(rsslState.DataState());
            m_State.Code(rsslState.Code());
            m_State.Text(rsslState.Text());
        }
        else
        {
            Code = DataCode.BLANK;

            m_State.Code(Eta.Codec.StateCodes.NONE);
            m_State.DataState(Eta.Codec.DataStates.OK);
            m_State.StreamState(Eta.Codec.StreamStates.OPEN);
            m_State.Text().Clear();
        }
    }

    internal override string ToString(int indent)
    {
        return ToString();
    }

    /* This is used to override the data state from the ETA State */
    internal void SetDataState(int dataState)
    {
        m_State.DataState(dataState);
    }

    private LSEG.Eta.Codec.State m_State = new();

    internal const string OPEN_STRING = "Open";
    internal const string NONSTREAMING_STRING = "NonStreaming";
    internal const string CLOSED_STRING = "Closed";
    internal const string CLOSEDRECOVER_STRING = "ClosedRecover";
    internal const string CLOSEDREDIRECTED_STRING = "ClosedRedirected";

    internal const string NOCHANGE_STRING = "NoChange";
    internal const string OK_STRING = "Ok";
    internal const string SUSPECT_STRING = "Suspect";

    internal const string NONE_STRING = "None";
    internal const string NOTFOUND_STRING = "NotFound";
    internal const string TIMEOUT_STRING = "Timeout";
    internal const string NOTAUTHORIZED_STRING = "NotAuthorized";
    internal const string INVALIDARGUMENT_STRING = "InvalidArgument";
    internal const string USAGEERROR_STRING = "UsageError";
    internal const string PREEMPTED_STRING = "Preempted";
    internal const string JUSTINTIMECONFLATIONSTARTED_STRING = "JustInTimeConflationStarted";
    internal const string TICKBYTICKRESUMED_STRING = "TickByTickResumed";
    internal const string FAILOVERSTARTED_STRING = "FailoverStarted";
    internal const string FAILOVERCOMPLETED_STRING = "FailoverCompleted";
    internal const string GAPDETECTED_STRING = "GapDetected";
    internal const string NORESOURCES_STRING = "NoResources";
    internal const string TOOMANYITEMS_STRING = "TooManyItems";
    internal const string ALREADYOPEN_STRING = "AlreadyOpen";
    internal const string SOURCEUNKNOWN_STRING = "SourceUnknown";
    internal const string NOTOPEN_STRING = "NotOpen";
    internal const string NONUPDATINGITEM_STRING = "NonUpdatingItem";
    internal const string UNSUPPORTEDVIEWTYPE_STRING = "UnsupportedViewType";
    internal const string INVALIDVIEW_STRING = "InvalidView";
    internal const string FULLVIEWPROVIDED_STRING = "FullViewProvided";
    internal const string UNABLETOREQUESTASBATCH_STRING = "UnableToRequestAsBatch";
    internal const string NOBATCHVIEWSUPPORTINREQ_STRING = "NoBatchViewSupportInReq";
    internal const string EXCEEDEDMAXMOUNTSPERUSER_STRING = "ExceededMaxMountsPerUser";
    internal const string ERROR_STRING = "Error";
    internal const string DACSDOWN_STRING = "DacsDown";
    internal const string USERUNKNOWNTOPERMSYS_STRING = "UserUnknownToPermSys";
    internal const string DACSMAXLOGINSREACHED_STRING = "DacsMaxLoginsReached";
    internal const string DACSUSERACCESSTOAPPDENIED_STRING = "DacsUserAccessToAppDenied";
    internal const string GAPFILL_STRING = "GapFill";
    internal const string APPAUTHORIZATIONFAILED_STRING = "AppAuthorizationFailed";

    private const string DEFAULTSS_STRING = "Unknown StreamState value ";
    private const string DEFAULTDS_STRING = "Unknown DataState value ";
    private const string DEFAULTSC_STRING = "Unknown StatusCode value ";

    #endregion
}
